### PR TITLE
feat(app): add jupyter advanced setting to download offset data

### PR DESCRIPTION
**Pull request by sakibh**
# Overview

This PR adds a toggle in the advanced settings for jupyter and ssh users to download offset data. This addresses the first task in #9059

image

# Changelog

- Added toggle item in advanced settings
- Added selector for labware offset data toggle
- Added new schema type in redux config
- Added test coverage for selector and component

# Review requests

- Check that the toggle item matches the design and copy of the [Figma design](https://www.figma.com/file/JCQNOTTK9tTmYPhRqeOfSj/Milestone-0%3A-Protocol-Upload-Revamp?node-id=9846%3A186829)
- Check that the selector is implemented and working properly for use in downstream tasks

# Risk assessment

low

### CODE REVIEW

File: app/src/redux/config/schema-types.ts [hunk 1]
```diff
@@ -105,6 +105,9 @@ export interface ConfigV3 extends Omit<ConfigV2, 'version' | 'support'> {
     name: string | null
     email: string | null
   }
+  robotSettings: {
+    labwareOffsetData: boolean
+  }
```
Comment (sakibh) [inline app/src/redux/config/schema-types.ts]: Not sure what the naming should be for the schema in redux and also the name of the selector. Open to suggestions.

**Review by shlokamin — CHANGES_REQUESTED** [review]
I think you're almost there! I'm getting a white screen when i navigate to the more tab where this FF toggle lives. I suspect this is because we need to add a config migration in `app-shell/src/config/migrate.ts` bcuz right now no `robotSettings` property exists on the config. 

lets talk about this after standup though and we can resolve this!

File: app-shell/src/config/migrate.ts [hunk 2]
```diff
@@ -129,25 +128,39 @@ const toVersion3 = (prevConfig: ConfigV2): ConfigV3 => {
...
-  let result: ConfigV0 | ConfigV1 | ConfigV2 | ConfigV3 = prevConfig
+  let result = prevConfig
 
   // loop through the migrations, skipping any migrations that are unnecessary
-  for (let i = prevVersion; i < MIGRATIONS.length; i++) {
-    // NOTE(mc, 2020-05-22): Flow will always be unable to resolve this type
-    // ensure good unit test coverage for this logic
+  for (let i: number = prevVersion; i < MIGRATIONS.length; i++) {
     const migrateVersion = MIGRATIONS[i]
+    // @ts-expect-error (sh: 01-24-2021): migrateVersion function input typed to never
```
Comment (shlokamin) [inline app-shell/src/config/migrate.ts:163]: Anyone know why this is happening? I couldn't figure it out :/

File: app-shell/src/buildroot/index.ts [hunk 3]
```diff
@@ -88,14 +88,14 @@ export function registerBuildrootUpdate(dispatch: Dispatch): Dispatch {
         // eslint-disable-next-line @typescript-eslint/no-floating-promises
         uploadSystemFile(host, path, file)
           .then(() => ({
-            type: 'buildroot:FILE_UPLOAD_DONE',
+            type: 'buildroot:FILE_UPLOAD_DONE' as 'buildroot:FILE_UPLOAD_DONE',
```
Comment (shlokamin) [inline app-shell/src/buildroot/index.ts]: my bad for not thinking of this earlier, but these `as` casts can all be changed to `as const` instead which is more accurate and safer

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions
```suggestion
            type: 'buildroot:FILE_UPLOAD_DONE' as const,
```

File: app-shell/src/buildroot/index.ts [hunk 4]
```diff
@@ -88,14 +88,14 @@ export function registerBuildrootUpdate(dispatch: Dispatch): Dispatch {
...
             payload: host.name,
           }))
           .catch((error: Error) => {
             log.warn('Error uploading update to robot', { path, file, error })
 
             return {
-              type: 'buildroot:UNEXPECTED_ERROR',
+              type: 'buildroot:UNEXPECTED_ERROR' as 'buildroot:UNEXPECTED_ERROR',
```
Comment (shlokamin) [inline app-shell/src/buildroot/index.ts]:
```suggestion
              type: 'buildroot:UNEXPECTED_ERROR' as const,
```

File: app-shell/src/buildroot/index.ts [hunk 5]
```diff
@@ -112,14 +112,14 @@ export function registerBuildrootUpdate(dispatch: Dispatch): Dispatch {
         // eslint-disable-next-line @typescript-eslint/no-floating-promises
         readUserFileInfo(systemFile)
           .then(userFile => ({
-            type: 'buildroot:USER_FILE_INFO',
+            type: 'buildroot:USER_FILE_INFO' as 'buildroot:USER_FILE_INFO',
```
Comment (shlokamin) [inline app-shell/src/buildroot/index.ts]:
```suggestion
            type: 'buildroot:USER_FILE_INFO' as const,
```

File: app-shell/src/buildroot/index.ts [hunk 6]
```diff
@@ -112,14 +112,14 @@ export function registerBuildrootUpdate(dispatch: Dispatch): Dispatch {
...
             payload: {
               systemFile: userFile.systemFile,
               version: userFile.versionInfo.opentrons_api_version,
             },
           }))
           .catch((error: Error) => ({
-            type: 'buildroot:UNEXPECTED_ERROR',
+            type: 'buildroot:UNEXPECTED_ERROR' as 'buildroot:UNEXPECTED_ERROR',
```
Comment (shlokamin) [inline app-shell/src/buildroot/index.ts]:
```suggestion
            type: 'buildroot:UNEXPECTED_ERROR' as const,
```

**Review by shlokamin — APPROVED** [review]
left some suggestions, but code changes look good to me and I tested that the config correctly toggles in redux!

File: app-shell/src/config/migrate.ts (hunk 2 above)
Comment (shlokamin) [reply app-shell/src/config/migrate.ts:163]: @b-cooper @mcous

Comment (b-cooper) [reply app-shell/src/config/migrate.ts:163]: """
multiple candidates for the same type variable in contra-variant positions causes an intersection type to be inferred.
"""
It looks like we can avoid this with a slight refactor of `MIGRATIONS` const by making it an object (then we loose ordinality). Probably best to punt that refactor to another day though

**Review by b-cooper — APPROVED** [review]

**Pull request merged**